putty.go

```diff
@@ -519,34 +729,81 @@ func (k Key) deriveKeys(password []byte) ([]byte, []byte, []byte, error) {
 		nil
 }
 
-// decrypt decrypts the key, when it is encrypted. and validates its signature
-func (k *Key) decrypt(password []byte) error {
+// Decrypt decrypts the key, when it is encrypted. and validates its signature
+func (k *Key) decrypt(password []byte) (err error) {
 	cipherKey, cipherIV, macKey, err := k.deriveKeys(password)
 	if err != nil {
 		return err
 	}
 
 	// decrypt the key, when it is encrypted
-	if !k.decrypted && k.Encryption != "none" {
+	if k.Encryption != "none" {
 		err = decryptCBC(cipherKey, cipherIV, macKey, k.PrivateKey)
 		if err != nil {
 			return err
 		}
 	}
-	k.decrypted = true
 
 	// validate key signature
 	switch k.Version {
+	case 1:
+		h := sha1.New().Sum(k.PrivateKey)
+		if !bytes.Equal(h, k.PrivateMac) {
+			return fmt.Errorf("calculated HASH %q doesn't correspond to %q", hex.EncodeToString(h), hex.EncodeToString(k.PrivateMac))
+		}
+		return nil
 	case 2:
 		err = k.validateHMAC(hmac.New(sha1.New, macKey))
 	case 3:
 		err = k.validateHMAC(hmac.New(sha256.New, macKey))
 	default:
 		err = fmt.Errorf("unknown key format version: %d", k.Version)
 	}
+	if err != nil {
+		return
+	}
+
+	return
+}
+
+// Encrypt encrypts the key and updates the HMAC
+func (k *Key) Encrypt(random io.Reader, password []byte) error {
+	if k.Encryption != "none" {
+		return fmt.Errorf("First decrypt the key before you can encrypt it")
```

Comment:
```suggestion
		return fmt.Errorf("first decrypt the key before you can encrypt it")
```

Reply:
Maybe "First decrypt before calling encrypt"?

Reply:
Or cannot encrypt a key as it is already encrypted